File: irrgarten-java/src/irrgarten/Dice.java
package irrgarten;

import java.util.Random;

public class Dice {
    private static final int MAX_USES = 5;
    private static final float MAX_INTELLIGENCE = 10f;
    private static final float MAX_STRENGTH = 10f;
    private static final float RESURRECT_PROB = .3f;
    private static final int WEAPONS_REWARD = 2;
    private static final int SHIELDS_REWARD = 3;
    private static final int HEALTH_REWARD = 5;
    private static final float MAX_ATTACK = 3f;
    private static final float MAX_SHIELD = 2f;

    private static Random generator = new Random();
 
    public static int randomPos(int max) {
        return generator.nextInt(max + 1);
    }
    
    public static int whoStarts(int nPlayers) {
        return generator.nextInt(nPlayers);
    }
    
    public static float randomIntelligence() {
        return generator.nextFloat(MAX_INTELLIGENCE);
    }
    
    public static float randomStrength() {
        return generator.nextFloat(MAX_STRENGTH);
    }
    
    public static boolean resurrectPlayer() {
        return generator.nextFloat() <= RESURRECT_PROB;
    }
    
    public static int weaponsReward() {
        return generator.nextInt(WEAPONS_REWARD + 1);
    }
    
    public static int shieldsReward() {
        return generator.nextInt(SHIELDS_REWARD + 1);
    }
    
    public static int healthReward() {
        return generator.nextInt(HEALTH_REWARD + 1);
    }
    
    public static float weaponPower() {
        return generator.nextFloat(MAX_ATTACK);
    }
    
    public static float shieldPower() {
        return generator.nextFloat(MAX_SHIELD);
    }
    
    public static int usesLeft() {
        return generator.nextInt(MAX_USES + 1);
    }
    
    public static float intensity(float competence) {
        return generator.nextFloat(competence);
    }
    
    public static boolean discardElement(int usesLeft) {
        return generator.nextFloat(MAX_USES) > usesLeft;
    }
}

File: irrgarten-ruby/dice.rb
# frozen_string_literal: true
module Irrgarten

  class Dice
    @@MAX_USES = 5
    @@MAX_INTELLIGENCE = 10.0
    @@MAX_STRENGTH = 10.0
    @@RESURRECT_PROB = 0.3
    @@WEAPONS_REWARD = 2
    @@SHIELDS_REWARD = 3
    @@HEALTH_REWARD = 5
    @@MAX_ATTACK = 3
    @@MAX_SHIELD = 2

    @@generator = Random.new

    def self.random_pos(max)
      @@generator.rand(max)
    end

    def self.who_starts(nplayers)
      @@generator.rand(nplayers)
    end

    def self.random_intelligence
      @@generator.rand(@@MAX_USES)
    end

    def self.random_strength
      @@generator.rand(@@MAX_STRENGTH)
    end

    def self.resurrect_player
      @@generator.rand(1.0) <= @@RESURRECT_PROB
    end

    def self.weapons_reward
      @@generator.rand(@@WEAPONS_REWARD+1)
    end

    def self.shields_reward
      @@generator.rand(@@SHIELDS_REWARD+1)
    end

    def self.health_reward
      @@generator.rand(@@HEALTH_REWARD+1)
    end

    def self.weapon_power
      @@generator.rand(@@MAX_ATTACK)
    end

    def self.shield_power
      @@generator.rand(@@MAX_SHIELD)
    end

    def self.uses_left
      @@generator.rand(@@MAX_USES+1)
    end

    def self.intensity(competence)
      @@generator.rand(competence)
    end

    def self.discard_element(uses_left)
      # @@generator.rand devuelve un float entre 0 y 1
      @@generator.rand < (@@MAX_USES.to_f - uses_left) / @@MAX_USES.to_f
    end
  end

end
